stdlib/public/Concurrency/TaskCancellation.swift

```diff
@@ -0,0 +1,169 @@
+  /// by invoking `Task.checkCancellation`, or `Task.isCancelled`.
+  ///
+  /// - Parameters:
+  ///   - interval: interval after which (from `now()`) the operation task should 
+  ///     be considered cancelled.
+  ///   - operation: the operation to execute
+  public static func withDeadline<T>(
+    in interval: _TimeInterval,
+    operation: () async throws -> T
+  ) async rethrows -> T {
+    fatalError("\(#function) not implemented yet.")
+  }
+
+  /// Execute a code block with the passed in deadline (unless a shorter deadline is already set).
+  ///
+  /// If the current task already has a deadline set that is _prior_
+  /// to the newly set deadline with this API, that deadline will remain in effect.
+  ///
+  /// This allows higher level tasks to set an upper bound on the deadline they
+  /// are willing cumulatively willing to wait for the entire task to execute,
+  /// regardless of the inner deadlines of the specific child tasks.
+  ///
+  /// Cancellation is co-operative and must be checked for by the operation, e.g.
+  /// by invoking `Task.checkCancellation` or `Task.isCancelled`.
+  ///
+  /// - Parameters:
+  ///   - deadline: the point in time after which the operation task should be
+  ///     considered cancelled.
+  ///   - operation: the operation to execute
+  public static func withDeadline<T>(
+    _ deadline: Deadline,
+    operation: () async throws -> T
+  ) async rethrows -> T {
+    fatalError("\(#function) not implemented yet.")
+  }
+
+  /// A deadline is a point in time past-which a task should be considered cancelled.
+  ///
+  /// Deadlines function the same was as pure cancellation, in the sense that they
+  /// are cooperative and require the cancelled (deadline exceeding) task to check
+  /// for this as it is performing its execution.
+  ///
+  /// Generally tasks (or partial tasks) should perform such check before they
+  /// start executing, however this is not a strict rule, and some tasks may
+  /// choose to be un-cancellable.
+  public struct Deadline {
+    typealias WallTime = UInt64 // equivalent to DispatchWallTime
+    internal let time: WallTime
+
+    public static var distantFuture: Self {
+      .init(time: .max)
+    }
+
+    public static func `in`(_ interval: _TimeInterval) -> Self {
+      // now() + interval
+      fatalError("#\(#function) not implemented yet.")
+    }
+
+    /// Returns `true` if the deadline is overdue and deadline should be
+    /// considered overdue (or "exceeded").
+    ///
+    /// If this deadline was related to a `Task`, that task should be considered
+    /// cancelled if the deadline is overdue.
+    public func isOverdue() -> Bool {
+      !self.hasTimeLeft()
+    }
+
+    /// Returns `true` if the deadline is still pending with respect to "now".
+    public func hasTimeLeft() -> Bool {
+      fatalError("\(#function) not implemented yet.")// self.hasTimeLeft(until: now())
+    }
```

Comment:
Maybe we can use properties instead of methods to simplify the call sites.
```suggestion
    /// A Boolean value indicating whether the deadline is overdue and the
    /// deadline should be considered overdue (or "exceeded").
    ///
    /// If this deadline was related to a `Task`, that task should be considered
    /// cancelled if the deadline is overdue.
    public var isOverdue: Bool {
      !self.hasTimeLeft
    }

    /// A Boolean value indicating whether the deadline is still pending with
    /// respect to "now".
    public var hasTimeLeft: Bool {
      fatalError("\(#function) not implemented yet.")// self.hasTimeLeft(until: now())
    }
```

Reply:
Yeah, those can be properties, thanks 👍